# **INDIAN STARTUP ECOSYSTEM**

## Project Description
We embark on a journey of discovery as we leverage our data analysis expertise to uncover the untapped potential within the Indian startup ecosystem. This project is designed to not only decode the numbers but to distill insights that will guide our team towards a successful foray into this dynamic market.

## Scope of Work

- Conduct a thorough exploration of datasets, dissecting funding patterns, sectoral nuances, and geographical hotspots in the Indian startup landscap
- Analyze funding received by startups in india from 2018 to 2021



## Hypothesis 

**Null Hypothesis (H0)**: There is no significant relationship between funding and the sector  

**Alternative Hypothesis (H1)**: There is a significant relationship between funding and the sector

## Questions 
1. How does funding vary across different industry sectors in India?
2. How does funding vary with the loaction of the start-ups
3. What is the relationship between the amount of funding and the stage of the company?
4. How have funding trends evolved between 2018 and 2021?
5. What are the most attractive sectors for investors?
6. Does the location of the company influence its sector?




# **DATA EXPLORATION, DATA UNDERSTANDING and DATA ANALYSIS**

In [1]:
# Load libraries
# Database connnection
import pyodbc     
from dotenv import dotenv_values    #import the dotenv_values function from the dotenv package

# Data manipulation
import pandas as pd
import numpy as np

# Data visualization
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

# filter warnings
import warnings 
warnings.filterwarnings('ignore')

# **1. Loading and Inspection of Data**

## **1.1 Loading data from the SQL server**

In [2]:
# Load environment variables from .env file into a dictionary
environment_variables = dotenv_values('.env')

# Get the values for the credentials you set in the '.env' file
server = environment_variables.get("SERVER")
database = environment_variables.get("DATABASE")
username = environment_variables.get("USERNAME")
password = environment_variables.get("PASSWORD")

In [3]:
# Create a connection string

connection_string = f"DRIVER={{SQL Server}};SERVER={server};DATABASE={database};UID={username};PWD={password}"
    
connection = pyodbc.connect(connection_string)



In [4]:
# sql query to get 2020 data. 
query_2020="SELECT * FROM dbo.LP1_startup_funding2020"

# sql query to get 2021 data. 
query_2021="SELECT * FROM dbo.LP1_startup_funding2021"

In [5]:
    # load 2021 data
data_2021=pd.read_sql(query_2021,connection)

    # load 2020 data
data_2020=pd.read_sql(query_2020,connection)

## **1.2 Loading CSV Files**

In [6]:
# load 2019 data
data_2019=pd.read_csv(r'C:\Users\iamde\OneDrive\Desktop\jupyter\india_startup_data\startup_funding2019.csv')

    # load 2018 data
data_2018=pd.read_csv(r'C:\Users\iamde\OneDrive\Desktop\jupyter\india_startup_data\startup_funding2018.csv')


# **2.Exploratory Data Analaysis(EDA)**

## **2.0.  2018 Dataset EDA**

In [7]:
# preview the rows and columns for the 2018 dataset
data_2018.sample(5)

Company Name                                           Industry  \
228    MY CHIRAAG CAB                                             Travel   
179         Colive.in                                           Consumer   
368  Idha Skin Clinic   Beauty, Cosmetics, Health Care, Service Industry   
455            Disprz                              E-Learning, Education   
464     Stanza Living  Hospitality, Information Technology, Travel Ac...   

                 Round/Series        Amount                          Location  \
228                      Seed       1000000          Jaipur, Rajasthan, India   
179                      Seed       1800000       Bangalore, Karnataka, India   
368                      Seed    ₹5,000,000  Bangalore City, Karnataka, India   
455            Debt Financing        400000        Chennai, Tamil Nadu, India   
464  Venture - Series Unknown  ₹730,000,000           New Delhi, Delhi, India   

                                         About Company  
228                    MY CHIRAAG CAB SERVICES PVT LTD  
179  A technology powered network of Branded Servic...  
368  Idha skin clinic jayanagar is dedicated to res...  
455  Disprz is a one-of-its-kind “knowledge-tech” s...  
464  Stanza Living is a co-living platform created ...

In [8]:
# checking for number of columns and rows
print (data_2018.shape)
print(f"There are {data_2018.shape[0]} rows, and {data_2018.shape[1]} columns")

(526, 6)
There are 526 rows, and 6 columns


In [9]:
# checking info
data_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 526 entries, 0 to 525
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Company Name   526 non-null    object
 1   Industry       526 non-null    object
 2   Round/Series   526 non-null    object
 3   Amount         526 non-null    object
 4   Location       526 non-null    object
 5   About Company  526 non-null    object
dtypes: object(6)
memory usage: 24.8+ KB


In [10]:
# Describing the data
data_2018.describe()

Company Name Industry Round/Series Amount  \
count               526      526          526    526   
unique              525      405           21    198   
top     TheCollegeFever        —         Seed      —   
freq                  2       30          280    148   

                           Location  \
count                           526   
unique                           50   
top     Bangalore, Karnataka, India   
freq                            102   

                                            About Company  
count                                                 526  
unique                                                524  
top     TheCollegeFever is a hub for fun, fiesta and f...  
freq                                                    2

**Findings**  
- TheCollegeFever company was the most common company  

- Seed series was the most preffered  

- most companies were based in Bangalore location  

In [11]:
# checking for duplicates
print("There are ",data_2018.duplicated().sum(),"duplicate(s)")

There are  1 duplicate(s)


In [12]:
# Checking for missing values
data_2018.isnull().sum()

Company Name     0
Industry         0
Round/Series     0
Amount           0
Location         0
About Company    0
dtype: int64

## **2.1. 2019 EDA**

In [13]:
# get a sample of 2019 dataset
data_2019.sample(5)

Company/Brand  Founded HeadQuarter      Sector  \
8         Quantiphi      NaN         NaN   AI & Tech   
71  Smart Institute      NaN       Surat      Edtech   
55      Binca Games   2014.0      Mumbai       Games   
74            Credr   2014.0      Mumbai  Automobile   
60     Ecom Express   2012.0    Gurugram   Logistics   

                                         What it does  \
8   It is an AI and big data services company prov...   
71                         A platform to learn online   
55  It offers games that are available across Indi...   
74  A marketplace to buy and sell used bikes and s...   
60  It is a technology enabled logistics solutions...   

                                             Founders  \
8                                      Renuka Ramnath   
71                                                NaN   
55                      Rubianca Wadhwa, Sahil Wadhwa   
74                                             Mittal   
60  Krishnan, Manju Dhawan, K Satyanarayana, Sanje...   

                                             Investor    Amount($)     Stage  
8                Multiples Alternate Asset Management  $20,000,000  Series A  
71                                      HNI investors   $1,500,000       NaN  
55                                     Sunil Gavaskar  Undisclosed       NaN  
74  Omidyar Network, Eight Roads Ventures, GrowthS...   $6,000,000       NaN  
60                                          CDC Group  $36,000,000       NaN

In [14]:
# checking for number of columns and rows
print (data_2019.shape)
print(f"There are {data_2019.shape[0]} rows, and {data_2019.shape[1]} columns")

(89, 9)
There are 89 rows, and 9 columns


In [15]:
# checking for duplicates
print("There are ",data_2019.duplicated().sum(),"duplicate(s)")

There are  0 duplicate(s)


In [16]:
# Checking for nulls
data_2019.isnull().sum()

Company/Brand     0
Founded          29
HeadQuarter      19
Sector            5
What it does      0
Founders          3
Investor          0
Amount($)         0
Stage            46
dtype: int64

In [17]:
# checking for datatypes in the different columns
data_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89 entries, 0 to 88
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Company/Brand  89 non-null     object 
 1   Founded        60 non-null     float64
 2   HeadQuarter    70 non-null     object 
 3   Sector         84 non-null     object 
 4   What it does   89 non-null     object 
 5   Founders       86 non-null     object 
 6   Investor       89 non-null     object 
 7   Amount($)      89 non-null     object 
 8   Stage          43 non-null     object 
dtypes: float64(1), object(8)
memory usage: 6.4+ KB


In [18]:
# performing descriptive analysis
data_2019.describe(include='all')

Company/Brand      Founded HeadQuarter  Sector      What it does  \
count             89    60.000000          70      84                89   
unique            87          NaN          17      52                88   
top         Kratikal          NaN   Bangalore  Edtech  Online meat shop   
freq               2          NaN          21       7                 2   
mean             NaN  2014.533333         NaN     NaN               NaN   
std              NaN     2.937003         NaN     NaN               NaN   
min              NaN  2004.000000         NaN     NaN               NaN   
25%              NaN  2013.000000         NaN     NaN               NaN   
50%              NaN  2015.000000         NaN     NaN               NaN   
75%              NaN  2016.250000         NaN     NaN               NaN   
max              NaN  2019.000000         NaN     NaN               NaN   

                          Founders     Investor    Amount($)     Stage  
count                           86           89           89        43  
unique                          85           86           50        15  
top     Vivek Gupta, Abhay Hanjura  Undisclosed  Undisclosed  Series A  
freq                             2            3           12        10  
mean                           NaN          NaN          NaN       NaN  
std                            NaN          NaN          NaN       NaN  
min                            NaN          NaN          NaN       NaN  
25%                            NaN          NaN          NaN       NaN  
50%                            NaN          NaN          NaN       NaN  
75%                            NaN          NaN          NaN       NaN  
max                            NaN          NaN          NaN       NaN

**Findings**
- Kratikal company was the most common company

- Most companies had their headquarters based in Bangalore

- Majority of the companies are involved in the Edtech sector

## **2.2. 2020 EDA**

In [19]:
# get a sample of 2020 dataset
data_2020.sample(5)

Company_Brand  Founded HeadQuarter         Sector  \
848     Multi Living      NaN      Mumbai    Real Estate   
573          Openapp      NaN   Bangalore            IoT   
938            Qyuki      NaN      Mumbai          Media   
233   Genius Teacher   2016.0      Mumbai         Edtech   
1025            Rheo      NaN        None  Entertainment   

                                           What_it_does  \
848   It curates apartments from the homeowners, per...   
573                          Security solutions startup   
938                     Multi-channel broadcast network   
233   Genius Teacher creates engaging educational co...   
1025   Online e-sports streaming and community platform   

                            Founders  \
848                     Pankaj Singh   
573                     Gotama Gowda   
938         AR Rahman, Shekhar Kapur   
233   Advitiya Sharma, Gaurav Lahoti   
1025                   Gaurav Munjal   

                                        Investor     Amount Stage column10  
848                               Lodha Ventures  6000000.0  None     None  
573                       Unicorn India Ventures   500000.0  None     None  
938                       Info Edge Venture Fund  3500000.0  None     None  
233        Whiteboard Capital, VKG Ventures LLP.  2000000.0  None     None  
1025  Lightspeed Venture Partners, Sequoia Surge  2000000.0  None     None

In [20]:
#checking for number of columns and rows
print (data_2020.shape)
print(f"There are {data_2020.shape[0]} rows, and {data_2020.shape[1]} columns")

(1055, 10)
There are 1055 rows, and 10 columns


In [21]:
# checking for duplicates
print("There are ",data_2020.duplicated().sum(),"duplicate(s)")

There are  3 duplicate(s)


In [22]:
# Checking for nulls
data_2020.isnull().sum()

Company_Brand       0
Founded           213
HeadQuarter        94
Sector             13
What_it_does        0
Founders           12
Investor           38
Amount            254
Stage             464
column10         1053
dtype: int64

In [23]:
# checking for datatypes in the different columns
data_2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1055 entries, 0 to 1054
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Company_Brand  1055 non-null   object 
 1   Founded        842 non-null    float64
 2   HeadQuarter    961 non-null    object 
 3   Sector         1042 non-null   object 
 4   What_it_does   1055 non-null   object 
 5   Founders       1043 non-null   object 
 6   Investor       1017 non-null   object 
 7   Amount         801 non-null    float64
 8   Stage          591 non-null    object 
 9   column10       2 non-null      object 
dtypes: float64(2), object(8)
memory usage: 82.6+ KB


In [24]:
# Describe the data
data_2020.describe(include='all')

Company_Brand      Founded HeadQuarter   Sector  \
count           1055   842.000000         961     1042   
unique           905          NaN          77      302   
top            Nykaa          NaN   Bangalore  Fintech   
freq               6          NaN         317       80   
mean             NaN  2015.363420         NaN      NaN   
std              NaN     4.097909         NaN      NaN   
min              NaN  1973.000000         NaN      NaN   
25%              NaN  2014.000000         NaN      NaN   
50%              NaN  2016.000000         NaN      NaN   
75%              NaN  2018.000000         NaN      NaN   
max              NaN  2020.000000         NaN      NaN   

                            What_it_does       Founders           Investor  \
count                               1055           1043               1017   
unique                               990            927                848   
top     Provides online learning classes  Falguni Nayar  Venture Catalysts   
freq                                   4              6                 20   
mean                                 NaN            NaN                NaN   
std                                  NaN            NaN                NaN   
min                                  NaN            NaN                NaN   
25%                                  NaN            NaN                NaN   
50%                                  NaN            NaN                NaN   
75%                                  NaN            NaN                NaN   
max                                  NaN            NaN                NaN   

              Amount     Stage  column10  
count   8.010000e+02       591         2  
unique           NaN        42         2  
top              NaN  Series A  Pre-Seed  
freq             NaN        96         1  
mean    1.130430e+08       NaN       NaN  
std     2.476635e+09       NaN       NaN  
min     1.270000e+04       NaN       NaN  
25%     1.000000e+06       NaN       NaN  
50%     3.000000e+06       NaN       NaN  
75%     1.100000e+07       NaN       NaN  
max     7.000000e+10       NaN       NaN

**Findings**  
- Nykaa company was the most common company  

- Majority of the companies had their headquarter at Bangalore  

- The average Amount invested was 113042969.54

## **2.3. 2021 EDA**

In [25]:
# get a sample of 2021 dataset
data_2021.sample(5)

Company_Brand  Founded HeadQuarter             Sector  \
966          Trica   2021.0      Mumbai  Equity Management   
796      DealShare   2018.0      Jaipur         E-commerce   
759        MyGlamm   2015.0      Mumbai    Beauty products   
1118    Expertrons   2019.0      Mumbai         E-learning   
1046     Instoried   2018.0   Bangalore  Computer Software   

                                           What_it_does  \
966   Powering private markets - A technology platfo...   
796              DealShare is a Social Commerce Startup   
759   MyGlamm is a direct-to-consumer brand of award...   
1118  The world's largest AI Videobot platform provi...   
1046  The platform helps content writers in e-commer...   

                                       Founders  \
966                              Nimesh Kampani   
796   Sankar Bora, Sourjyendu Medda, Vineet Rao   
759                              Darpan Sanghvi   
1118                 Jatin Solanki, Vivek Gupta   
1046                                Sharmin Ali   

                                  Investor        Amount     Stage  
966                            LetsVenture      $3000000      Seed  
796   Omidyar Network, Falcon Edge Capital    $3,000,000      Debt  
759              Bessemer Venture Partners   $20,000,000  Series C  
1118          Kunal Shah, Anant Maheshwari  $Undisclosed      None  
1046  Pritt Investment Partners, 9Unicorns      $8000000      None

In [26]:
# get the number of rows and columns for the datasets
print(f'{data_2021.shape}')

print(f"The 2021 dataset has {data_2021.shape[0]} rows and {data_2021.shape[1]} Columns")

(1209, 9)
The 2021 dataset has 1209 rows and 9 Columns


In [27]:
# checking for duplicates
print("There are ",data_2021.duplicated().sum(),"duplicate(s)")

There are  19 duplicate(s)


In [28]:
#  Checking for nulls
data_2021.isnull().sum()

Company_Brand      0
Founded            1
HeadQuarter        1
Sector             0
What_it_does       0
Founders           4
Investor          62
Amount             3
Stage            428
dtype: int64

In [29]:
# checking for datatypes in the different columns
data_2021.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1209 entries, 0 to 1208
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Company_Brand  1209 non-null   object 
 1   Founded        1208 non-null   float64
 2   HeadQuarter    1208 non-null   object 
 3   Sector         1209 non-null   object 
 4   What_it_does   1209 non-null   object 
 5   Founders       1205 non-null   object 
 6   Investor       1147 non-null   object 
 7   Amount         1206 non-null   object 
 8   Stage          781 non-null    object 
dtypes: float64(1), object(8)
memory usage: 85.1+ KB


In [30]:
# Describe the dataset
data_2021.describe(include='all')

Company_Brand      Founded HeadQuarter   Sector  \
count           1209  1208.000000        1208     1209   
unique          1033          NaN          70      254   
top         BharatPe          NaN   Bangalore  FinTech   
freq               8          NaN         426      122   
mean             NaN  2016.655629         NaN      NaN   
std              NaN     4.517364         NaN      NaN   
min              NaN  1963.000000         NaN      NaN   
25%              NaN  2015.000000         NaN      NaN   
50%              NaN  2018.000000         NaN      NaN   
75%              NaN  2020.000000         NaN      NaN   
max              NaN  2021.000000         NaN      NaN   

                                             What_it_does  \
count                                                1209   
unique                                               1143   
top     BharatPe develops a QR code-based payment app ...   
freq                                                    4   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                Founders                   Investor  \
count                               1205                       1147   
unique                              1095                        937   
top     Ashneer Grover, Shashvat Nakrani  Inflection Point Ventures   
freq                                   7                         24   
mean                                 NaN                        NaN   
std                                  NaN                        NaN   
min                                  NaN                        NaN   
25%                                  NaN                        NaN   
50%                                  NaN                        NaN   
75%                                  NaN                        NaN   
max                                  NaN                        NaN   

              Amount Stage  
count           1206   781  
unique           278    31  
top     $Undisclosed  Seed  
freq              73   246  
mean             NaN   NaN  
std              NaN   NaN  
min              NaN   NaN  
25%              NaN   NaN  
50%              NaN   NaN  
75%              NaN   NaN  
max              NaN   NaN

**Findings**  
- BharatPe is the most common company  

- Most companies invested in the FinTech sector  

- Majority of the companies were found in Bangalore  


**Overall Observations:**  
**Issues with the data**

1. There is a discrepancy in the naming conventions between the columns in the 2018 and 2019 datasets compared to the 2020 and 2021 datasets.

2. The 2018 dataset exhibits some missing columns, contributing to an incomplete representation of the data.

3. Conversely, the 2020 dataset contains an additional column that appears to be extraneous and does not serve a meaningful purpose in our analysis.

**Course of Action:**

1. **Missing Column Engineering for 2018:**
   - We will address the absence of certain columns in the 2018 dataset by employing data engineering techniques to create and populate the missing columns, ensuring a comprehensive and consistent dataset.

2. **Column Name Standardization:**
   - To establish uniformity and coherence across all datasets, we will embark on a column renaming process for the 2018 and 2019 datasets. This action aims to align the naming conventions with those observed in the 2020 and 2021 datasets, facilitating seamless data integration and analysis.

3. **Extraneous Column Removal in 2020:**
   - The redundant column identified in the 2020 dataset will be removed, streamlining the dataset and eliminating unnecessary elements that do not contribute to the overall analysis objectives.

These actions collectively enhance the integrity, consistency, and completeness of the dataset, paving the way for a more robust and coherent analytical process.
ical process.







# **3. Data Cleaning and Preparation**

## **3.1. Convert Indian Rupee values in the 2018 data to dollars**


In [31]:
# Replace '₹', commas, '—', and "''" in 'Amount' column
data_2018['Amount'] = data_2018['Amount'].str.replace(',', '').str.replace('—', '').str.replace("''",'').replace('', np.nan)

# Conditionally apply multiplication only where '₹' is present
mask = data_2018['Amount'].str.contains('₹', na=False)
data_2018.loc[mask, 'Amount'] = data_2018.loc[mask, 'Amount'].str.replace('₹', '').astype(float) * 0.0146

### **Add year data was collected**

- There is need to add a column that represents the year each dataset was collected. This will help with handling the datasets after merging the dataframes 


In [32]:
# add year when data was collected column to every dataset
data_2018['year'] = pd.to_datetime('2018', format='%Y').year
data_2019['year'] = pd.to_datetime('2019', format='%Y').year
data_2020['year'] = pd.to_datetime('2020', format='%Y').year
data_2021['year'] = pd.to_datetime('2021', format='%Y').year

In [33]:
def concat_dataframes(data,data1, data2, data3):
    """
    Concatenates four dataframes along the row axis.

    Parameters:
    data (pandas.DataFrame): The first dataframe to concatenate.
    data1 (pandas.DataFrame): The second dataframe to concatenate.
    data2 (pandas.DataFrame): The third dataframe to concatenate.
    data3 (pandas.DataFrame): The fourth dataframe to concatenate.

    Returns:
    pandas.DataFrame: The concatenated dataframe.
    """
    # Rename columns to make them consistent across dataframes
    df1 = data.rename(columns={'Company_Brand': 'company_brand', 'Founded': 'founded','HeadQuarter': 'headquarter','Sector': 'sector','What_it_does': 'about_company','Founders':'founders','Investor':'investor','Amount':'amount','Stage':'series', 'year' : 'year'})
    df2 = data1.rename(columns={'Company_Brand': 'company_brand', 'Founded': 'founded','HeadQuarter': 'headquarter','Sector': 'sector','What_it_does': 'about_company','Founders':'founders','Investor':'investor','Amount':'amount','Stage':'series','year' : 'year'})
    df3 = data2.rename(columns={'Company/Brand': 'company_brand', 'Founded': 'founded','HeadQuarter': 'headquarter','Sector': 'sector','What it does': 'about_company','Founders':'founders','Investor':'investor','Amount($)':'amount','Stage':'series','year' : 'year'})
    df4 = data3.rename(columns={'Company Name': 'company_brand','Location': 'headquarter','Industry': 'sector','About Company': 'about_company','Amount':'amount','Round/Series':'series','year' : 'year'})

    # Concatenate dataframes along the row axis
    result = pd.concat([df1, df2, df3,df4], axis=0).reset_index(drop=True)

    return result

## **Concatenate the dataframes**

In [34]:
# concating the dataframes
df= concat_dataframes(data_2021,data_2020,data_2019,data_2018)

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2879 entries, 0 to 2878
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   company_brand  2879 non-null   object 
 1   founded        2110 non-null   float64
 2   headquarter    2765 non-null   object 
 3   sector         2861 non-null   object 
 4   about_company  2879 non-null   object 
 5   founders       2334 non-null   object 
 6   investor       2253 non-null   object 
 7   amount         2474 non-null   object 
 8   series         1941 non-null   object 
 9   year           2879 non-null   int64  
 10  column10       2 non-null      object 
dtypes: float64(1), int64(1), object(9)
memory usage: 247.5+ KB


In [36]:
#Drop the extreneous column 10
df.drop('column10', axis=1, inplace= True)

## **Cleaning 'Amount' column**

**Notes**  
- Remove all currency signs  

- Remove all other umwanted characters, words and symbols  

- Fill the nulls uning interpolate method 

- Convert the column from object to float

In [37]:
df['amount']=df['amount'].replace('\$','',regex=True)
df['amount']=df['amount'].replace(['Undisclosed','undisclosed',None],np.NAN)
df['headquarter']=df['headquarter'].str.strip('\t#REF!')
df['amount']=df['amount'].replace('\,','',regex=True)
df['amount']=df['amount'].replace(",","")



In [38]:
# fix index 242
df.at[242, 'headquarter'] = None
df.at[242, 'sector'] = 'Pharmaceuticals'
df.at[242, 'founders'] = 'Varun Khanna'
df.at[242, 'investor'] = 'Morgan Stanley Private Equity Asia'
df.at[242, 'amount'] = '22000000'
df.at[242, 'series'] = 'Series C'
df.at[242, 'about_company'] = 'Development and Manufacturing'

# drop the duplicate row
df = df.drop(256, axis=0, errors='ignore')

# fix index 1148
df.at[1148, 'sector'] = 'Logistics'
df.at[1148, 'founders'] ='Basant Kumar' 
df.at[1148, 'investor'] =None
df.at[1148, 'amount'] = 1000000
df.at[1148, 'series'] = 'Seed'
 
# fix index 257
df.at[257, 'sector'] = 'Transport'
df.at[257, 'founders'] ='Anshuman Maheshwary'
df.at[257, 'investor'] =None
df.at[257, 'amount'] = 5000000
df.at[257, 'series'] = 'Seed'

# fix index 98
df.at[98, 'sector'] = 'Computer Games'
df.at[98, 'founders'] ='Pritesh Kumar'
df.at[98, 'investor'] =None
df.at[98, 'amount'] = 1200000
df.at[98, 'series'] = 'Seed'

# drop the duplicate column
df = df.drop(111, axis=0, errors='ignore')

# fix index 677
df.at[677, 'amount'] = 1000000
df.at[677, 'series'] = 'Angel'

# fix index 545
df.at[545, 'investor'] =None
df.at[545, 'amount'] = 1000000
df.at[545, 'series'] = 'Pre-series A'

# fix index 538
df.at[538, 'investor'] =None
df.at[538, 'amount'] = 300000
df.at[538, 'series'] = 'Pre-Seed'


In [39]:
df.head()

company_brand  founded headquarter          sector  \
0  Unbox Robotics   2019.0   Bangalore      AI startup   
1          upGrad   2015.0      Mumbai          EdTech   
2     Lead School   2012.0      Mumbai          EdTech   
3         Bizongo   2015.0      Mumbai  B2B E-commerce   
4        FypMoney   2021.0    Gurugram         FinTech   

                                       about_company  \
0  Unbox Robotics builds on-demand AI-driven ware...   
1     UpGrad is an online higher education platform.   
2  LEAD School offers technology based school tra...   
3  Bizongo is a business-to-business online marke...   
4  FypMoney is Digital NEO Bank for Teenagers, em...   

                                            founders  \
0                        Pramod Ghadge, Shahid Memon   
1  Mayank Kumar, Phalgun Kompalli, Ravijot Chugh,...   
2                         Smita Deorah, Sumeet Mehta   
3            Aniket Deb, Ankit Tomar, Sachin Agrawal   
4                                      Kapil Banwari   

                                        investor     amount        series  \
0                    BEENEXT, Entrepreneur First    1200000  Pre-series A   
1       Unilazer Ventures, IIFL Asset Management  120000000          None   
2               GSV Ventures, Westbridge Capital   30000000      Series D   
3                         CDC Group, IDG Capital   51000000      Series C   
4  Liberatha Kallat, Mukesh Yadav, Dinesh Nagpal    2000000          Seed   

   year  
0  2021  
1  2021  
2  2021  
3  2021  
4  2021

## **Cleaning data_year column**

**Notes**  


- Convert data type to period


In [41]:
# Convert the data_year column to date
df['year']=pd.to_datetime(df['year'], format='%Y')
df['year']=df['year'].dt.to_period('y')
df['founded']=pd.to_datetime(df['founded']).dt.year

In [44]:
# check for nulls and duplicated
print(f"There are {df['year'].isna().sum()} Null values in the 'year' column")

There are 0 Null values in the 'year' column


In [45]:
df['year'].info()

<class 'pandas.core.series.Series'>
Index: 2877 entries, 0 to 2878
Series name: year
Non-Null Count  Dtype        
--------------  -----        
2877 non-null   period[Y-DEC]
dtypes: period[Y-DEC](1)
memory usage: 109.5 KB


## **Cleaning 'founded' column**

**Notes**
- Handle nulls by populating with the 'bfill' method



In [46]:
print(f"There are {df['founded'].isna().sum()} Null values in the 'founded' column")

There are 769 Null values in the 'founded' column


**Notes**  
- There are 769 null values in the 'founded' column.  

- Since dropping the nulls will lead to a significant loss of our data, Backward fill will be used to fill the null values

**COURSE OF ACTION**

- We will be cleaning the missing values by fill them using interpolate method which is suitable for time series data  

- We will also be converting the data type from float to datetime for purpose of our analysis.

In [47]:
# Fill the nulls
df["founded"].interpolate(method='linear',inplace= True)

print(f"There are {df['founded'].isna().sum()} missing values")

There are 0 missing values


In [48]:
# Convert to datetime
df['founded'] = pd.to_datetime(df['founded'], format='%Y')

# Convert to period
df['founded'] = df['founded'].dt.to_period('Y')


In [49]:
df['founded'].info()

<class 'pandas.core.series.Series'>
Index: 2877 entries, 0 to 2878
Series name: founded
Non-Null Count  Dtype        
--------------  -----        
2877 non-null   period[Y-DEC]
dtypes: period[Y-DEC](1)
memory usage: 109.5 KB


In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2877 entries, 0 to 2878
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype        
---  ------         --------------  -----        
 0   company_brand  2877 non-null   object       
 1   founded        2877 non-null   period[Y-DEC]
 2   headquarter    2762 non-null   object       
 3   sector         2859 non-null   object       
 4   about_company  2877 non-null   object       
 5   founders       2332 non-null   object       
 6   investor       2246 non-null   object       
 7   amount         2322 non-null   object       
 8   series         1944 non-null   object       
 9   year           2877 non-null   period[Y-DEC]
dtypes: object(8), period[Y-DEC](2)
memory usage: 311.8+ KB


## **Cleaning the 'founders' column**

In [51]:
# Remove unwanted characters
df['founders'] = df['founders'].replace(['...', np.nan], np.NaN)

# Check the number of NaN values in the 'founders' column
nan_count = df['founders'].isna().sum()

print(nan_count)

545


In [52]:
df['founders'].info()

<class 'pandas.core.series.Series'>
Index: 2877 entries, 0 to 2878
Series name: founders
Non-Null Count  Dtype 
--------------  ----- 
2332 non-null   object
dtypes: object(1)
memory usage: 109.5+ KB


## **Cleaning **Stage** column**

Startups start with pre-seed, progress through seed, Series A, Series B, etc., securing resources for development and strategies. Additional rounds like Series C or D may follow. External funding at each stage fuels growth toward the venture's full potential.

**Pre-Seed Funding**  
Entrepreneurial idea in early development; small funds needed; limited informal channels for raising funds.

**Seed Funding**  
First official equity funding; investors provide funds for equity ownership.

**Series A Financing**  
First venture capital round; developed product, consistent revenue, long-term profit plan.

**Series B Financing**  
For established startups; substantial user base and revenue; funding for expansion.

**Series C and Beyond**  
Optional rounds for final push before IPO or unmet objectives; Series C is the third venture capital round.

**Initial Public Offering (IPO)**  
Process of offering corporate shares to the public; used for funding or divestment.

link: https://www.startupindia.gov.in/content/sih/en/funding.html

In [56]:
df[df['series']=='$6000000']

company_brand founded headquarter                  sector  \
674  MYRE Capital    1970      Mumbai  Commercial Real Estate   

                           about_company  \
674  Democratising Real Estate Ownership   

                                    founders     investor amount    series  \
674  Own rent yielding commercial properties  Aryaman Vir    NaN  $6000000   

     year  
674  2021

In [ ]:
df.at()

In [57]:
# Cleaning stage column
df['series'].unique()

array(['Pre-series A', None, 'Series D', 'Series C', 'Seed', 'Series B',
       'Series E', 'Pre-seed', 'Series A', 'Pre-series B', 'Debt',
       'Bridge', 'Seed+', 'Series F2', 'Series A+', 'Series G',
       'Series F', 'Series H', 'Series B3', 'PE', 'Series F1',
       'Pre-series A1', 'Pre-Seed', '$300000', 'Early seed', 'Series D1',
       '$6000000', 'Angel', 'Seies A', 'Pre-series', 'Series A2',
       'Series I', 'Pre-series C', 'Series B2', 'Pre- series A', 'Edge',
       'Pre-Series B', 'Seed A', 'Series A-1', 'Seed Funding',
       'Seed round', 'Pre-seed Round', 'Seed Round & Series A',
       'Pre Series A', 'Pre seed Round', 'Angel Round', 'Pre series A1',
       'Series E2', 'Pre series A', 'Seed Round', 'Bridge Round',
       'Pre seed round', 'Pre series B', 'Pre series C',
       'Seed Investment', 'Mid series', 'Series C, D', 'Seed funding',
       nan, 'Fresh funding', 'Post series A', 'Seed fund', 'Series B+',
       'Private Equity', 'Venture - Series Unknown', '

In [54]:
# Standardize funding stages in the 'stage' column
df['stage'] = df['stage'].replace(['Series A', 'Seies A', 'Series A-1', 'Series A2', 'Series A+', 'Series A+'], 'Series A')
df['stage'] = df['stage'].replace(['Pre-seed', 'Pre-seed Round', 'Pre seed Round', 'Pre seed round'], 'Pre-Seed Stage')
df['stage'] = df['stage'].replace(['Pre series A', 'Pre-series A', 'Pre Series A', 'Pre series A1', 'Pre-series A1', 'Pre- series A'], 'Pre series A')
df['stage'] = df['stage'].replace(['Series B', 'Series B+', 'Series B2', 'Series B3'], 'Series B')
df['stage'] = df['stage'].replace(['Series C', 'Series C', 'Series C, D','Series C', 'Private Equity','PE', 'Post-IPO Equity','Series D', 'Series E', 'Series F', 'Series G', 'Series H', 'Series I','Series D1','Series F2', 'Series F1'], 'Series C and Beyond')
df['stage'] = df['stage'].replace(['Venture - Series Unknown', None,'Grant','Debt','Debt Financing','Post-IPO Debt','Non-equity Assistance','Bridge','Bridge Round','Fresh funding','Funding Round','Mid series','Edge',], 'unknown')
df['stage'] = df['stage'].replace(['Corporate Round','Undisclosed','Secondary Market','Pre-series','Post series A','Pre-series B','Pre-Series B','Pre series B','Pre-series C','Pre series C'], 'Other Stages')
df['stage'] = df['stage'].replace(['Seed','Seed funding','Pre-Seed','Angel', 'Angel Round','Seed fund', 'Seed round', 'Seed A','Seed Funding', 'Seed Round & Series A', 'Series E2', 'Seed Round','Seed Investment','Seed+','Early seed'],'Seed Stage')

KeyError: 'stage'

In [55]:
# strip off characters 
df['stage'] = df['stage'].str.strip('\t#REF!')

KeyError: 'stage'

## **Cleaning the Sector Column**

In [ ]:
# Get the first sentence of every list
df['sector']=df['sector'].str.split(",").str[0]

In [ ]:
df['sector'] = df['sector'].replace({
    'Edtech': 'EdTech', 'Fintech': 'FinTech', 'Agriculture': 'AgriTech', 'Food & Beverages': 'Food and Beverages',
    'Financial Services': 'FinTech', 'Healthcare': 'HealthTech', 'HealthTech': 'HealthTech', 'Medical': 'HealthTech', 
    'Medtech': 'HealthTech', 'Pharmaceutical': 'HealthTech', 'Health Insurance': 'HealthTech', 
    'Biotechnology': 'HealthTech', 'Health Diagnostics': 'HealthTech', 'Hospital': 'HealthTech', 
    'Hospital & Health Care': 'HealthTech', 'Wellness': 'HealthTech', 'Dental': 'HealthTech', 
    'Alternative Medicine': 'HealthTech', 'Nutrition': 'HealthTech', 'Fitness': 'HealthTech', 
    'Mental Health': 'HealthTech', 'Healthcare/Edtech': 'HealthTech',
    'Life sciences': 'HealthTech', 'Biotech': 'Healthcare', 'Nutrition Tech': 'HealthTech', 
    'E-mobility': 'HealthTech', 'Med Tech': 'HealthTech', 'FemTech': 'HealthTech', 
    'Cannabis startup': 'HealthTech', 'Pharmacy': 'HealthTech', 'Medical Device': 'HealthTech', 
    'BioTechnology': 'HealthTech', 'Fertility tech': 'HealthTech', 'Ayurveda tech': 'HealthTech', 
    'E-tail': 'Healthcare', 'E store': 'E-Commerce', 'E-store': 'Healthcare', 'Telemedicine': 'HealthTech', 
    'HealthCare': 'HealthTech', 'AI startup': 'AI', 'Information Services': 'InfoTech & Services', 
    'Healthtech': 'HealthTech', 'Finance': 'FinTech', 'Health Care': 'HealthTech', 
    'Logistics & Supply Chain': 'Logistics', 'Food Industry': 'FoodTech', 'Foodtech': 'FoodTech', 
    '—': 'Undisclosed', 'SaaS startup': 'SaaS', 'Health': 'HealthTech', 'Ecommerce': 'E-Commerce', 
    'Tech Startup': 'Tech', 'Mobility': 'Transportation', 'SaaS': 'Tech', 'Artificial Intelligence': 'AI', 
    'Food and Beverage': 'Food and Beverages', 'Information Technology': 'InfoTech & Services', 
    'Internet': 'Tech', 'Apps': 'Tech', 'Computer Software': 'Tech', 'E-commerce': 'E-Commerce', 
    'Agritech': 'AgriTech', 'Hospital & Health Care': 'HealthTech', 'Food': 'Foodtech', 'Cosmetics': 'Consumer Goods', 
    'Tech company': 'Tech', 'Automobile': 'Automotive', 'Apparel & Fashion': 'Fashion', 'Education': 'EdTech', 
    'Social Media': 'Media', 'Digital Media': 'Media', 'IT': 'InfoTech & Services', 'IoT': 'AI', 
    'Software': 'Tech', 'Industrial Automation': 'AI', 'Technology': 'Tech', 
    'Information Technology & Services': 'InfoTech & Services', None: 'Unknown'
})




## **Cleaning HeadQuater Column**

In [ ]:
# Get the first location from every list
df['headquater']=df['headquater'].str.split(",").str[0]

In [ ]:
df['headquater']=df['headquater'].replace({'Bengaluru': 'Bangalore', 'Banglore': 'Bangalore', 'Gurugram': 'Gurgaon', 'Hyderebad': 'Hyderabad', 
                                      'New Delhi': 'Delhi', 'Ahmadabad': 'Ahmedabad', 'Ernakulam': 'Cochin', 'Telugana': 'Telangana',
                                      'Rajastan': 'Rajasthan', 'San Franciscao': 'San Francisco', 'Samsitpur': 'Samastipur', 'Santra': 'Samtra',
                                      'Rajsamand': 'Rajasthan', 'Kerala': 'Kochi','The Nilgiris': 'Nilgiris', 'Gurugram\t#REF!': 'Gurgaon', 
                                      'California': 'San Francisco', 'San Francisco Bay Area': 'San Francisco', 'Hyderebad': 'Hyderabad',
                                      'Online Media\t#REF!': 'Unknown','Pharmaceuticals\t#REF!': 'Unknown',
                                      'Information Technology & Services':'Unknown' ,'Small Towns': 'Unknown','Orissia': 'Odisha', 
                                      'Santra': 'Samtra', 'Vadodara': 'Vadodara', 'Food & Beverages': 'Unknown', 'Bangaldesh': 'Bangladesh',}) 


In [ ]:
df.head()

company_brand                sector       stage  amount($) headquater  \
0  TheCollegeFever       Brand Marketing  Seed Stage   250000.0  Bangalore   
1  Happy Cow Dairy              AgriTech  Seed Stage        NaN     Mumbai   
2       MyLoanCare                Credit    Series A        NaN    Gurgaon   
3      PayMe India               FinTech  Seed Stage  2000000.0      Noida   
4         Eunimart  E-Commerce Platforms  Seed Stage        NaN  Hyderabad   

                                       about_company founded founders  \
0  TheCollegeFever is a hub for fun, fiesta and f...     NaT      NaN   
1  A startup which aggregates milk from dairy far...     NaT      NaN   
2          Leading Online Loans Marketplace in India     NaT      NaN   
3  PayMe India is an innovative FinTech organizat...     NaT      NaN   
4  Eunimart is a one stop solution for merchants ...     NaT      NaN   

  investor data_year  
0      NaN      2018  
1      NaN      2018  
2      NaN      2018  
3      NaN      2018  
4      NaN      2018

In [ ]:
# check for duplicates
df[df.duplicated()]

company_brand                                             sector  \
348      TheCollegeFever                                    Brand Marketing   
760            Krimanshi                              Biotechnology company   
820                Nykaa                                     Consumer Goods   
977               Byju’s                                             EdTech   
982               Zomato                                     Food devlivery   
1428               Nykaa                                         E-Commerce   
1578                Vogo                                         Automotive   
1624              Bounce                             Automotive and Rentals   
1777           Curefoods                                 Food and Beverages   
1779            Bewakoof                                            Fashion   
1781             FanPlay                                     Computer Games   
1787      Advantage Club                                             HRTech   
1789              Ruptok                                            FinTech   
1913            Trinkerr                                    Capital Markets   
1914               Zorro                                     Social network   
1915       Ultraviolette                                         Automotive   
1916          NephroPlus                                         HealthTech   
1917             Unremot                                InfoTech & Services   
1918         FanAnywhere                                            FinTech   
1919          PingoLearn                                         E-learning   
1920                Spry                                              Music   
1921             Enmovil                                InfoTech & Services   
1922       ASQI Advisors                                            FinTech   
1923  Insurance Samadhan                                          Insurance   
1924     Evenflow Brands                                     Consumer Goods   
1925          MasterChow                                          Hauz Khas   
1926  Fullife Healthcare  Primary Business is Development and Manufactur...   

                    stage    amount($)      headquater  \
348            Seed Stage     250000.0       Bangalore   
760            Seed Stage          NaN         Jodhpur   
820               unknown          NaN          Mumbai   
977               unknown          NaN       Bangalore   
982               unknown          NaN         Haryana   
1428              unknown          NaN          Mumbai   
1578  Series C and Beyond          NaN       Bangalore   
1624  Series C and Beyond          NaN       Bangalore   
1777              unknown   13000000.0       Bangalore   
1779              unknown    8000000.0          Mumbai   
1781              unknown          NaN  Computer Games   
1787              unknown    1700000.0          Mumbai   
1789              unknown    1000000.0           Delhi   
1913             Series A    6600000.0       Bangalore   
1914           Seed Stage   32000000.0         Gurgaon   
1915  Series C and Beyond  150000000.0       Bangalore   
1916  Series C and Beyond   24000000.0       Hyderabad   
1917           Seed Stage     700000.0       Bangalore   
1918           Seed Stage          NaN       Bangalore   
1919              unknown     200000.0            Pune   
1920           Seed Stage    1500000.0          Mumbai   
1921         Pre series A    1300000.0       Hyderabad   
1922         Pre series A    1000000.0          Mumbai   
1923         Pre series A     725000.0           Delhi   
1924              unknown     500000.0          Mumbai   
1925           Seed Stage     461000.0         Unknown   
1926              unknown          NaN         Unknown   

                                          about_company founded  \
348   TheCollegeFever is a hub for fun, fiesta and f...     NaT   
760   Krimanshi aims to increase rural 

In [ ]:
df.drop_duplicates(keep='first', inplace= True)
df.head()

company_brand                sector       stage  amount($) headquater  \
0  TheCollegeFever       Brand Marketing  Seed Stage   250000.0  Bangalore   
1  Happy Cow Dairy              AgriTech  Seed Stage        NaN     Mumbai   
2       MyLoanCare                Credit    Series A        NaN    Gurgaon   
3      PayMe India               FinTech  Seed Stage  2000000.0      Noida   
4         Eunimart  E-Commerce Platforms  Seed Stage        NaN  Hyderabad   

                                       about_company founded founders  \
0  TheCollegeFever is a hub for fun, fiesta and f...     NaT      NaN   
1  A startup which aggregates milk from dairy far...     NaT      NaN   
2          Leading Online Loans Marketplace in India     NaT      NaN   
3  PayMe India is an innovative FinTech organizat...     NaT      NaN   
4  Eunimart is a one stop solution for merchants ...     NaT      NaN   

  investor data_year  
0      NaN      2018  
1      NaN      2018  
2      NaN      2018  
3      NaN      2018  
4      NaN      2018

In [ ]:
# check for duplicates
print(f" There are {df.duplicated().sum()} duplicates")

 There are 0 duplicates


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2852 entries, 0 to 2878
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype        
---  ------         --------------  -----        
 0   company_brand  2852 non-null   object       
 1   sector         2852 non-null   object       
 2   stage          2852 non-null   object       
 3   amount($)      1350 non-null   float64      
 4   headquater     2738 non-null   object       
 5   about_company  2852 non-null   object       
 6   founded        2326 non-null   period[Y-DEC]
 7   founders       2308 non-null   object       
 8   investor       2228 non-null   object       
 9   data_year      2852 non-null   period[Y-DEC]
dtypes: float64(1), object(7), period[Y-DEC](2)
memory usage: 245.1+ KB
